packages/core/src/utils/checkpoint.ts

```diff
@@ -55,77 +89,68 @@ export const decodeCheckpoint = (checkpoint: string): Checkpoint => {
   const blockNumber = +checkpoint.slice(offset, offset + BLOCK_NUMBER_DIGITS);
   offset += BLOCK_NUMBER_DIGITS;
 
-  let logIndex: number | undefined = Number(
-    checkpoint.slice(offset, offset + EXECUTION_INDEX_DIGITS),
+  const transactionIndex = +checkpoint.slice(
+    offset,
+    offset + TRANSACTION_INDEX_DIGITS,
   );
-
-  if (logIndex > Number.MAX_SAFE_INTEGER) logIndex = undefined;
-
-  return { blockTimestamp, chainId, blockNumber, logIndex };
+  offset += TRANSACTION_INDEX_DIGITS;
+
+  const eventType = +checkpoint.slice(offset, offset + EVENT_TYPE_DIGITS);
+  offset += EVENT_TYPE_DIGITS;
+
+  const eventIndex = +checkpoint.slice(offset, offset + EVENT_INDEX_DIGITS);
+  offset += EVENT_INDEX_DIGITS;
+
+  return {
+    blockTimestamp,
+    chainId,
+    blockNumber,
+    transactionIndex,
+    eventType,
+    eventIndex,
+  };
 };
 
 export const zeroCheckpoint: Checkpoint = {
   blockTimestamp: 0,
   chainId: 0,
   blockNumber: 0,
-  logIndex: 0,
+  transactionIndex: 0,
+  eventType: 0,
+  eventIndex: 0,
 };
 
 // TODO: Consider changing block timestamps and numbers to bigints
 // so that we can accurately represent EVM max values.
 export const maxCheckpoint: Checkpoint = {
   blockTimestamp: 9999999999,
-  chainId: 2147483647,
+  chainId: 9999999999,
   blockNumber: 9999999999,
-  logIndex: 2147483647,
+  transactionIndex: 9999999999,
+  eventType: 9,
+  eventIndex: 9999999999,
 };
 
 /**
  * Returns true if two checkpoints are equal.
  */
-export const isCheckpointEqual = (a: Checkpoint, b: Checkpoint) => {
-  return (
-    a.blockTimestamp === b.blockTimestamp &&
-    a.chainId === b.chainId &&
-    a.blockNumber === b.blockNumber &&
-    ((a.logIndex === undefined && b.logIndex === undefined) ||
-      a.logIndex === b.logIndex)
-  );
-};
+export const isCheckpointEqual = (a: Checkpoint, b: Checkpoint) =>
+  encodeCheckpoint(a) === encodeCheckpoint(b);
 
 /**
  * Returns true if checkpoint a is greater than checkpoint b.
  * Returns false if the checkpoints are equal.
  */
-export const isCheckpointGreaterThan = (a: Checkpoint, b: Checkpoint) => {
-  if (a.blockTimestamp !== b.blockTimestamp)
-    return a.blockTimestamp > b.blockTimestamp;
-  if (a.chainId !== b.chainId) return a.chainId > b.chainId;
-  if (a.blockNumber !== b.blockNumber) return a.blockNumber > b.blockNumber;
-
-  // If both logIndex are defined, compare normally.
-  if (a.logIndex !== undefined && b.logIndex !== undefined) {
-    return a.logIndex > b.logIndex;
-  }
-  // If both are undefined, the checkpoints are equal, so a is not greater than b.
-  if (a.logIndex === undefined && b.logIndex === undefined) {
-    return false;
-  }
-  // If only a is undefined, it's considered greater.
-  if (a.logIndex === undefined) return true;
-  // Otherwise b is undefined and a is defined, so b is greater.
-  else return false;
-};
+export const isCheckpointGreaterThan = (a: Checkpoint, b: Checkpoint) =>
+  encodeCheckpoint(a) > encodeCheckpoint(b);
```

Comment:
Any idea the difference in performance of this change? This function could be called millions of times.

Reply:
Prob a bit slower since its doing some allocation on the heap for the encoding bit. Makes the code much more readable and less error prone but fine to change it back to comparing every field manually if that's better. I don't think it'll make any noticeable diff in runtime perf but fine to change it back and if we'd prefer the manual one